File: 06-DevOps/07-DevSecOps/05-策略即代码Rego/python/safety.py
#!/usr/bin/env python3
"""OPA ``v1/ast`` 的 ``Unify`` 与 ``outputVarsForExpr``（逐行转写）。

* ``v1/ast/unify.go``   —— ``Unify`` / ``unifier`` / ``isRefSafe``
* ``v1/ast/compile.go`` —— ``outputVarsForExpr`` / ``outputVarsForExprEq`` /
  ``outputVarsForExprCall`` / ``outputVarsForTerms`` / ``outputVarsForBody``

两个极易写反的点：

* ``outputVarsForTerms`` 的结果要与 ``safe`` **取并集**再交给 ``Unify``，
  最后再减掉 ``safe``；不是「取 Unify 结果代替」。
* ``:=`` 会把 LHS 的变量**整体**从安全基里删掉（issue #3546），
  所以 ``input.a[i] == 1`` 产出 ``{i}`` 而 ``input.a[i] := 1`` 产出空集。
"""

from __future__ import annotations

from typing import Any, Iterator, List, Set, Tuple

from terms import Array, CallT, Const, Obj, Ref, SetT, Term, Var, is_ref_safe
from exprs import Bare, Call, Eq, Every, Expr, Logical, Not, With

# Unify（v1/ast/unify.go）
# --------------------------------------------------------------------------


class Unifier:
    def __init__(self, safe: Set[str]):
        self.safe = set(safe)
        self.unified: Set[str] = set()
        self.unknown: dict = {}

    def is_safe(self, v: str) -> bool:
        return v in self.safe or v in self.unified

    def mark_safe(self, v: str) -> None:
        self.unified.add(v)
        for w in list(self.unknown.pop(v, ())):
            self.mark_safe(w)
        for w in list(self.unknown.keys()):
            deps = self.unknown.get(w)
            if deps is None:
                continue
            if v in deps:
                deps.discard(v)
                if not deps:
                    del self.unknown[w]
                    self.mark_safe(w)

    def mark_unknown(self, a: str, b: str) -> None:
        self.unknown.setdefault(a, set()).add(b)

    def mark_all_safe(self, t: Term) -> None:
        for v in t.output_vars():
            self.mark_safe(v)

    def unify_all(self, a: Var, b: Term) -> None:
        if self.is_safe(a.name):
            self.mark_all_safe(b)
            return
        unsafe = b.output_vars() - self.safe - self.unified
        if not unsafe:
            self.mark_safe(a.name)
        else:
            for v in unsafe:
                self.mark_unknown(a.name, v)

    def unify(self, a: Term, b: Term) -> None:
        if isinstance(a, Var):
            if isinstance(b, Var):
                if self.is_safe(b.name):
                    self.mark_safe(a.name)
                elif self.is_safe(a.name):
                    self.mark_safe(b.name)
                else:
                    self.mark_unknown(a.name, b.name)
                    self.mark_unknown(b.name, a.name)
            elif isinstance(b, (Array, Obj)):
                self.unify_all(a, b)
            elif isinstance(b, Ref):
                if is_ref_safe(b, self.safe):
                    self.mark_safe(a.name)
            else:
                self.mark_safe(a.name)          # default: Const / Set / 调用项
        elif isinstance(a, Ref):
            if is_ref_safe(a, self.safe):
                if isinstance(b, Var):
                    self.mark_safe(b.name)
                elif isinstance(b, (Array, Obj)):
                    self.mark_all_safe(b)
        elif isinstance(a, Array):
            if isinstance(b, Var):
                self.unify_all(b, a)
            elif isinstance(b, Array) and len(a.items) == len(b.items):
                for x, y in zip(a.items, b.items):
                    self.unify(x, y)
        elif isinstance(a, Obj):
            if isinstance(b, Var):
                self.unify_all(b, a)
            elif isinstance(b, Obj) and len(a.pairs) == len(b.pairs):
                bm = {_okey(k): v for k, v in b.pairs}
                for k, v in a.pairs:
                    if _okey(k) in bm:
                        self.unify(v, bm[_okey(k)])
        else:                                    # default: Const / Set 在左侧
            if isinstance(b, Var):
                self.mark_safe(b.name)


def _okey(k: Term) -> Tuple[str, Any]:
    """对象键的比较口径：常量键按值，变量键按名字（OPA 用 Term.Get 精确匹配）。"""
    return ("c", k.value) if isinstance(k, Const) else ("t", repr(k))


def unify(a: Term, b: Term, safe: Set[str]) -> Set[str]:
    u = Unifier(safe)
    u.unify(a, b)
    return u.unified


# --------------------------------------------------------------------------
# outputVarsFor*
# --------------------------------------------------------------------------


def walk_terms(t: Term) -> Iterator[Term]:
    yield t
    if isinstance(t, Ref):
        yield from walk_terms(t.head)
        for p in t.parts:
            if isinstance(p, Term):
                yield from walk_terms(p)
    elif isinstance(t, Array):
        for i in t.items:
            yield from walk_terms(i)
    elif isinstance(t, Obj):
        for k, v in t.pairs:
            yield from walk_terms(k)
            yield from walk_terms(v)
    elif isinstance(t, SetT):
        for i in t.items:
            yield from walk_terms(i)
    elif isinstance(t, CallT):
        for i in t.operands:
            yield from walk_terms(i)


def output_vars_for_terms(e: Any, safe: Set[str]) -> Set[str]:
    """outputVarsForTerms：收集「ref-safe 且非 ground」的引用的下标变量。"""
    out: Set[str] = set()
    src = [e] if isinstance(e, Term) else list(e.terms())
    for t in src:
        for x in walk_terms(t):
            if isinstance(x, Ref):
                if not is_ref_safe(x, safe):
                    continue
                if not x.is_ground():
                    out |= x.output_vars()
    return out


def output_vars_for_eq(e: Eq, safe: Set[str]) -> Set[str]:
    out = output_vars_for_terms(e, safe)
    out |= set(safe)
    if e.from_assignment:
        # 剔除 LHS：防止 RHS 反向通过 LHS 获得安全性（issue #3546）
        out -= e.lhs.vars()
    u = Unifier(out)
    u.unify(e.lhs, e.rhs)
    out |= u.unified
    return out - set(safe)


def output_vars_for_call(e: Call, safe: Set[str]) -> Set[str]:
    if e.arity < 0:                      # 未知 arity（源码 ar < 0）
        return set()
    out = output_vars_for_terms(e, safe)
    # numInputTerms = arity + 1（含操作符项）⟺ arity >= len(operands) 时无输出项
    if e.arity >= len(e.operands):
        return out
    unsafe: Set[str] = set()
    for t in e.operands[:e.arity]:
        unsafe |= t.output_vars()
    unsafe -= out
    unsafe -= set(safe)
    if unsafe:
        return set()
    for t in e.operands[e.arity:]:
        out |= t.output_vars()
    return out


def output_vars_for_expr(e: Expr, safe: Set[str]) -> Set[str]:
    """outputVarsForExpr：这条表达式能产出哪些变量。"""
    if isinstance(e, Not):
        return set()                               # IsNegated
    if isinstance(e, With):
        if e.target.vars() - set(safe):            # with 输入必须已安全
            return set()
        return output_vars_for_expr(e.inner, safe)
    if isinstance(e, Bare):
        return output_vars_for_terms(e.term, safe)
    if isinstance(e, Every):
        return output_vars_for_terms(e.domain, safe)
    if isinstance(e, Logical):
        return set()                               # and / or 不贡献绑定
    if isinstance(e, Eq):
        return output_vars_for_eq(e, safe)
    if isinstance(e, Call):
        return output_vars_for_call(e, safe)
    return set()


def output_vars_for_body(body: List[Expr], safe: Set[str]) -> Set[str]:
    """outputVarsForBody：把 safe 逐条喂进去累积，最后再减掉原始 safe。"""
    o = set(safe)
    for e in body:
        o |= output_vars_for_expr(e, o)
    return o - set(safe)

File: 06-DevOps/07-DevSecOps/05-策略即代码Rego/go/safety.go
package main

// Unify（v1/ast/unify.go）与 outputVarsForExpr（v1/ast/compile.go）的转写。
//
// 两个极易写反的点：
//
//	outputVarsForTerms 的结果要与 safe 取并集再交给 Unify，最后再减掉 safe；
//	不是「拿 Unify 结果代替」。
//	:= 会把 LHS 的变量整体从安全基里删掉（issue #3546），所以
//	input.a[i] == 1 产出 {i} 而 input.a[i] := 1 产出空集。

// ---------------------------------------------------------------- Unifier

type Unifier struct {
	safe    VarSet
	unified VarSet
	unknown map[string]VarSet
}

func (u *Unifier) isSafe(v string) bool { return u.safe[v] || u.unified[v] }

func (u *Unifier) markSafe(v string) {
	u.unified[v] = true

	// x 的依赖随之安全
	deps := u.unknown[v]
	delete(u.unknown, v)
	for w := range deps {
		u.markSafe(w)
	}

	// 依赖 x 的变量若再无依赖，也变安全
	for w, ds := range u.unknown {
		if ds[v] {
			delete(ds, v)
			if len(ds) == 0 {
				delete(u.unknown, w)
				u.markSafe(w)
			}
		}
	}
}

func (u *Unifier) markUnknown(a, b string) {
	if u.unknown[a] == nil {
		u.unknown[a] = VarSet{}
	}
	u.unknown[a][b] = true
}

func (u *Unifier) markAllSafe(t Term) {
	for v := range t.OutputVars() {
		u.markSafe(v)
	}
}

func (u *Unifier) unifyAll(a *Var, b Term) {
	if u.isSafe(a.Name) {
		u.markAllSafe(b)
		return
	}
	unsafe := Diff(Diff(b.OutputVars(), u.safe), u.unified)
	if len(unsafe) == 0 {
		u.markSafe(a.Name)
		return
	}
	for v := range unsafe {
		u.markUnknown(a.Name, v)
	}
}

func (u *Unifier) unify(a, b Term) {
	switch x := a.(type) {

	case *Var:
		switch y := b.(type) {
		case *Var:
			if u.isSafe(y.Name) {
				u.markSafe(x.Name)
			} else if u.isSafe(x.Name) {
				u.markSafe(y.Name)
			} else {
				u.markUnknown(x.Name, y.Name)
				u.markUnknown(y.Name, x.Name)
			}
		case *Arr:
			u.unifyAll(x, y)
		case *Obj:
			u.unifyAll(x, y)
		case *Ref:
			if isRefSafe(y, u.safe) {
				u.markSafe(x.Name)
			}
		default: // Const / SetT / CallT
			u.markSafe(x.Name)
		}

	case *Ref:
		if isRefSafe(x, u.safe) {
			switch y := b.(type) {
			case *Var:
				u.markSafe(y.Name)
			case *Arr:
				u.markAllSafe(y)
			case *Obj:
				u.markAllSafe(y)
			}
		}

	case *Arr:
		switch y := b.(type) {
		case *Var:
			u.unifyAll(y, x)
		case *Arr:
			if len(x.Items) == len(y.Items) {
				for i := range x.Items {
					u.unify(x.Items[i], y.Items[i])
				}
			}
		}

	case *Obj:
		switch y := b.(type) {
		case *Var:
			u.unifyAll(y, x)
		case *Obj:
			if len(x.Keys) == len(y.Keys) {
				idx := map[string]Term{}
				for i := range y.Keys {
					idx[termKey(y.Keys[i])] = y.Vals[i]
				}
				for i := range x.Keys {
					if v, ok := idx[termKey(x.Keys[i])]; ok {
						u.unify(x.Vals[i], v)
					}
				}
			}
		}

	default: // Const / SetT 在左侧
		if y, ok := b.(*Var); ok {
			u.markSafe(y.Name)
		}
	}
}

// termKey 是对象键的比较口径：常量键按值，变量键按名字（OPA 用精确匹配）。
func termKey(t Term) string {
	if c, ok := t.(*Const); ok {
		return "c:" + c.Value
	}
	return "t:" + t.String()
}

func Unify(safe VarSet, a, b Term) VarSet {
	u := &Unifier{safe: safe, unified: VarSet{}, unknown: map[string]VarSet{}}
	u.unify(a, b)
	return u.unified
}

// ---------------------------------------------------------------- outputVars

func walkTerms(t Term, out *[]Term) {
	*out = append(*out, t)
	switch x := t.(type) {
	case *Ref:
		walkTerms(x.Head, out)
		for _, p := range x.Parts {
			walkTerms(p, out)
		}
	case *Arr:
		for _, i := range x.Items {
			walkTerms(i, out)
		}
	case *Obj:
		for i := range x.Keys {
			walkTerms(x.Keys[i], out)
			walkTerms(x.Vals[i], out)
		}
	case *SetT:
		for _, i := range x.Items {
			walkTerms(i, out)
		}
	case *CallT:
		for _, i := range x.Operands {
			walkTerms(i, out)
		}
	}
}

// outputVarsForTerms：收集「ref-safe 且非 ground」的引用的下标变量。
func outputVarsForTerms(e Expr, safe VarSet) VarSet {
	var all []Term
	if b, ok := e.(*Bare); ok {
		all = []Term{b.T}
	}
	for _, t := range e.Terms() {
		walkTerms(t, &all)
	}
	out := VarSet{}
	for _, x := range all {
		if r, ok := x.(*Ref); ok {
			if !isRefSafe(r, safe) {
				continue
			}
			if !r.IsGround() {
				out = Union(out, r.OutputVars())
			}
		}
	}
	return out
}

func outputVarsForEq(e *Eq, safe VarSet) VarSet {
	out := Union(outputVarsForTerms(e, safe), safe)
	if e.FromAssign {
		for v := range e.Lhs.Vars() { // issue #3546
			delete(out, v)
		}
	}
	out = Union(out, Unify(out, e.Lhs, e.Rhs))
	return Diff(out, safe)
}

func outputVarsForCall(e *Call, safe VarSet) VarSet {
	if e.Arity < 0 { // 未知 arity（源码 ar < 0）
		return VarSet{}
	}
	out := outputVarsForTerms(e, safe)
	// numInputTerms = arity + 1（含操作符项）
	if e.Arity >= len(e.Operands) {
		return out
	}
	unsafe := VarSet{}
	for _, t := range e.Operands[:e.Arity] {
		unsafe = Union(unsafe, t.OutputVars())
	}
	unsafe = Diff(Diff(unsafe, out), safe)
	if len(unsafe) > 0 {
		return VarSet{}
	}
	for _, t := range e.Operands[e.Arity:] {
		out = Union(out, t.OutputVars())
	}
	return out
}

func outputVarsForExpr(e Expr, safe VarSet) VarSet {
	switch x := e.(type) {
	case *Not: // IsNegated
		return VarSet{}
	case *With:
		if len(Diff(x.Target.Vars(), safe)) > 0 { // with 输入必须已安全
			return VarSet{}
		}
		return outputVarsForExpr(x.Inner, safe)
	case *Bare:
		return outputVarsForTerms(x, safe)
	case *Every:
		return outputVarsForTerms(x, safe)
	case *Logical: // and / or 不贡献绑定
		return VarSet{}
	case *Eq:
		return outputVarsForEq(x, safe)
	case *Call:
		return outputVarsForCall(x, safe)
	}
	return VarSet{}
}

// outputVarsForBody：把 safe 逐条喂进去累积，最后再减掉原始 safe。
func outputVarsForBody(body []Expr, safe VarSet) VarSet {
	o := Union(VarSet{}, safe)
	for _, e := range body {
		o = Union(o, outputVarsForExpr(e, o))
	}
	return Diff(o, safe)
}
